[LLVM][rtsan] Add RealtimeSanitizer transform pass

cjappl opened the pull request:
Split from #100596 at the request of the reviewers. (This review directly depends on that review being merged first)

Introduce the RealtimeSanitizer transform, which inserts the rtsan_enter/exit functions at the appropriate places in an instrumented function.

cjappl committed:
[LLVM][rtsan] Add RealtimeSanitizer transform pass

cjappl commented:
CC Reviewers @vitalybuka @MaskRay 

This review should be green once https://github.com/llvm/llvm-project/pull/100596 is merged, as this depends on the attribute existing

CC co-author @davidtrevelyan

vitalybuka commented on llvm/lib/Passes/PassRegistry.def:
```diff
@@ -590,6 +590,10 @@ FUNCTION_PASS_WITH_PARAMS(
     },
     parseWinEHPrepareOptions, "demote-catchswitch-only")
+FUNCTION_PASS_WITH_PARAMS(
+    "rtsan", "RealtimeSanitizerPass",
```
are these parameters expected in future?
If not, you can use simpler macro.

cjappl replied:
Yes, we are expecting to add options very soon, which is why I left it with the parameters (and the "empty" parameter file).

vitalybuka commented on llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:
```diff
@@ -0,0 +1,50 @@
+
+using namespace llvm;
+
+namespace {
```
I believe llvm prefer `static` instead of `namespace {}`

cjappl replied:
Fixed in latest

cjappl committed:
[PR] Add multi-return IR test

cjappl committed:
[PR] Ditch anonymous namespace

vitalybuka commented on llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:
```diff
@@ -0,0 +1,50 @@
+namespace {
+
+void insertCallBeforeInstruction(Function &Fn, Instruction &Instruction,
+                                 const char *FunctionName) {
```
Do you expect Module specific features? Other sanitizers are ModulePasses already.

cjappl replied:
At this time, no we do not expect to have Module specific features. A realtime or "nonblocking" context only deals with functions. We will insert appropriate calls at points within each function.

 As far as I can tell, the FunctionPass will be all that we need.

MaskRay replied:
Prefer static for internal linkage functions
https://llvm.org/docs/CodingStandards.html#anonymous-namespaces

cjappl commented:
Thanks for the approval and feedback. A reminder that this should not be merged before the attribute is settled in the other review (as I expect these tests to be red until then)

MaskRay commented on llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizerOptions.h:
```diff
@@ -0,0 +1,17 @@
+//===--------- Definition of the RealtimeSanitizer options -------*- C++
```
We don't add an unused header. This can be postponed to when the need arises.

MaskRay commented on llvm/test/Instrumentation/RealtimeSanitizer/rtsan.ll:
```diff
@@ -0,0 +1,35 @@
+attributes #4 = { nonblocking }
+
+; RealtimeSanitizer pass should insert __rtsan_realtime_enter right after function definition
+; CHECK: define{{.*}}violation
```
For functions, `@violation(` is preferred (also used by `llvm/utils/update_cc_test_checks.py`). This helps avoids ambiguity when you have `@foo` and `@foo2`

MaskRay commented on llvm/test/Instrumentation/RealtimeSanitizer/rtsan.ll:
```diff
@@ -0,0 +1,35 @@
+; RealtimeSanitizer pass should insert __rtsan_realtime_exit right before function return
+; CHECK: call{{.*}}@__rtsan_realtime_exit
+; CHECK-NEXT: ret{{.*}}void
+
```
no trailing blank lines

MaskRay commented on llvm/test/Instrumentation/RealtimeSanitizer/rtsan_multi_return.ll:
```diff
@@ -0,0 +1,29 @@
+
+attributes #0 = { mustprogress noinline nonblocking optnone ssp uwtable(sync) }
+
+; RealtimeSanitizer pass should insert __rtsan_realtime_enter right after function definition
```
prefer `CHECK-LABEL: @example(` in case more functions are added in the future.

MaskRay commented on llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizer.h:
```diff
@@ -0,0 +1,35 @@
+//===--------- Definition of the RealtimeSanitizer class ---------*- C++
```
Some `*Sanitizer.h` headers had non-conventional headers, which might lead to confusion. I fixed them in 
c6a3f4e2f2586cb1ce51306c305752a78bdba263

cjappl committed:
[PR] MaskRay feedback

cjappl commented:
Thanks @MaskRay ! Fixed all your comments in the latest version.

MaskRay commented on llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizer.h:
```diff
@@ -0,0 +1,34 @@
+//=- RealtimeSanitizer.h - RealtimeSanitizer instrumentation ----*- C++ -*-==//
+//-*-===//
```
unneeded  second line

the first line should use `//===-` (3 `=`). Both `//===-` and `//===-- ` are common in the code base and I am happy with either style.

cjappl replied:
Apologies, thanks for the double check! clang-format scrambled things a bit, should be fixed when I push this next version up.

MaskRay commented on llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizer.h:
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
```
Here a file-level comment is needed

MaskRay commented on llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:
```diff
@@ -0,0 +1,49 @@
+                                              const char *InsertFnName) {
+  for (auto &BB : Fn) {
+    for (auto &I : BB) {
+      if (auto *RI = dyn_cast<ReturnInst>(&I)) {
```
drop all the braces since the body is a simple statement with just one line

cjappl committed:
[PR] MaskRay feedback 2

cjappl commented:
Thanks for the reviews, now once the discussion on the attributes is finished in https://github.com/llvm/llvm-project/pull/100596 we can merge that, get green tests here and merge this.

I will ping this review again when that other one is completely settled. If anyone has any thoughts on the name of the attribute, please chime in over there.

cjappl committed:
Update attribute names to sanitize_realtime

cjappl commented:
@MaskRay @vitalybuka could I get a merge when you get a chance?

Since your last review I:
* Rebased on latest main
* changed the attribute name to be `sanitize_realtime` which was the change in my previous review. Everything else is the same as when you reviewed it.

Test being green shows that change took and everything is working as intended!

Thanks as always.

MaskRay commented on llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:
```diff
@@ -0,0 +1,46 @@
+#include "llvm/IR/Analysis.h"
```
This needs a file header

cjappl replied:
Fixed in latest!

cjappl committed:
Add file header for RealtimeSanitizer.cpp

cjappl commented:
Thanks for the help @MaskRay , always sincerely appreciated.

MaskRay commented:
> Thanks for the help @MaskRay , always sincerely appreciated.

Np:) Glad to help.

MaskRay commented:
> LLVM Buildbot has detected a new failure on builder `sanitizer-ppc64le-linux` running on `ppc64le-sanitizer` while building `llvm` at step 2 "annotate".
> 
> Full details are available at: [lab.llvm.org/buildbot#/builders/72/builds/2046](https://lab.llvm.org/buildbot/#/builders/72/builds/2046)
> 
> Here is the relevant piece of the build log for the reference:
> 
> ```
> Step 2 (annotate) failure: 'python ../sanitizer_buildbot/sanitizers/zorg/buildbot/builders/sanitizers/buildbot_selector.py' (failure)
> ...
> [3642/4007] Building CXX object tools/clang/tools/clang-import-test/CMakeFiles/clang-import-test.dir/clang-import-test.cpp.o
> [3643/4007] Building CXX object tools/clang/lib/Analysis/plugins/CheckerOptionHandling/CMakeFiles/CheckerOptionHandlingAnalyzerPlugin.dir/CheckerOptionHandling.cpp.o
> [3644/4007] Building CXX object tools/clang/tools/driver/CMakeFiles/clang.dir/driver.cpp.o
> [3645/4007] Building CXX object tools/clang/tools/clang-installapi/CMakeFiles/clang-installapi.dir/ClangInstallAPI.cpp.o
> [3646/4007] Building CXX object tools/clang/tools/clang-check/CMakeFiles/clang-check.dir/ClangCheck.cpp.o
> [3647/4007] Building CXX object tools/clang/tools/clang-extdef-mapping/CMakeFiles/clang-extdef-mapping.dir/ClangExtDefMapGen.cpp.o
> [3648/4007] Building CXX object tools/clang/tools/clang-repl/CMakeFiles/clang-repl.dir/ClangRepl.cpp.o
> [3649/4007] Building CXX object tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndex.cpp.o
> [3650/4007] Building CXX object tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXExtractAPI.cpp.o
> [3651/4007] Building CXX object lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o
> FAILED: lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o 
> CCACHE_CPP2=yes CCACHE_HASHDIR=yes /usr/bin/ccache /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm_build0/bin/clang++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -MF lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o.d -o lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -c /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp
> In file included from /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:20:
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizer.h:36:28: error: private field 'Options' is not used [-Werror,-Wunused-private-field]
>    36 |   RealtimeSanitizerOptions Options{};
>       |                            ^
> 1 error generated.
> [3652/4007] Building CXX object tools/clang/lib/Driver/CMakeFiles/obj.clangDriver.dir/ToolChains/HIPUtility.cpp.o
> [3653/4007] Building CXX object lib/MC/MCParser/CMakeFiles/LLVMMCParser.dir/AsmParser.cpp.o
> [3654/4007] Building CXX object lib/Passes/CMakeFiles/LLVMPasses.dir/PassBuilder.cpp.o
> ninja: build stopped: subcommand failed.
> + touch build_default/delete_next_time
> + build_failure
> + echo
> + echo 'How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild'
> + echo
> + sleep 5
> 
> How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild
> 
> + echo @@@STEP_FAILURE@@@
> + [[ '' == \1 ]]
> + [[ ! -v BUILDBOT_BUILDERNAME ]]
> + echo '@@@BUILD_STEP test compiler-rt debug@@@'
> + ninja -C build_default check-compiler-rt
> @@@STEP_FAILURE@@@
> @@@BUILD_STEP test compiler-rt debug@@@
> ninja: Entering directory `build_default'
> [1/238] Linking CXX static library lib/libLLVMMCParser.a
> [2/238] Linking CXX static library lib/libLLVMPowerPCAsmParser.a
> [3/238] Linking CXX static library lib/libLLVMObject.a
> [4/238] Linking CXX static library lib/libLLVMLibDriver.a
> [5/238] Linking CXX static library lib/libLLVMRuntimeDyld.a
> [6/238] Linking CXX static library lib/libLLVMDlltoolDriver.a
> [7/238] Linking CXX static library lib/libLLVMDebugInfoDWARF.a
> [8/238] Linking CXX static library lib/libLLVMObjCopy.a
> [9/238] Linking CXX static library lib/libLLVMObjectYAML.a
> [10/238] Linking CXX static library lib/libLLVMXRay.a
> [11/238] Linking CXX static library lib/libLLVMJITLink.a
> Step 8 (build compiler-rt debug) failure: build compiler-rt debug (failure)
> ...
> [3642/4007] Building CXX object tools/clang/tools/clang-import-test/CMakeFiles/clang-import-test.dir/clang-import-test.cpp.o
> [3643/4007] Building CXX object tools/clang/lib/Analysis/plugins/CheckerOptionHandling/CMakeFiles/CheckerOptionHandlingAnalyzerPlugin.dir/CheckerOptionHandling.cpp.o
> [3644/4007] Building CXX object tools/clang/tools/driver/CMakeFiles/clang.dir/driver.cpp.o
> [3645/4007] Building CXX object tools/clang/tools/clang-installapi/CMakeFiles/clang-installapi.dir/ClangInstallAPI.cpp.o
> [3646/4007] Building CXX object tools/clang/tools/clang-check/CMakeFiles/clang-check.dir/ClangCheck.cpp.o
> [3647/4007] Building CXX object tools/clang/tools/clang-extdef-mapping/CMakeFiles/clang-extdef-mapping.dir/ClangExtDefMapGen.cpp.o
> [3648/4007] Building CXX object tools/clang/tools/clang-repl/CMakeFiles/clang-repl.dir/ClangRepl.cpp.o
> [3649/4007] Building CXX object tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndex.cpp.o
> [3650/4007] Building CXX object tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXExtractAPI.cpp.o
> [3651/4007] Building CXX object lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o
> FAILED: lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o 
> CCACHE_CPP2=yes CCACHE_HASHDIR=yes /usr/bin/ccache /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm_build0/bin/clang++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -MF lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o.d -o lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -c /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp
> In file included from /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:20:
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizer.h:36:28: error: private field 'Options' is not used [-Werror,-Wunused-private-field]
>    36 |   RealtimeSanitizerOptions Options{};
>       |                            ^
> 1 error generated.
> [3652/4007] Building CXX object tools/clang/lib/Driver/CMakeFiles/obj.clangDriver.dir/ToolChains/HIPUtility.cpp.o
> [3653/4007] Building CXX object lib/MC/MCParser/CMakeFiles/LLVMMCParser.dir/AsmParser.cpp.o
> [3654/4007] Building CXX object lib/Passes/CMakeFiles/LLVMPasses.dir/PassBuilder.cpp.o
> ninja: build stopped: subcommand failed.
> + touch build_default/delete_next_time
> + build_failure
> + echo
> + echo 'How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild'
> + echo
> + sleep 5
> 
> How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild
> + echo @@@STEP_FAILURE@@@
> + [[ '' == \1 ]]
> + [[ ! -v BUILDBOT_BUILDERNAME ]]
> + echo '@@@BUILD_STEP test compiler-rt debug@@@'
> + ninja -C build_default check-compiler-rt
> Step 9 (test compiler-rt debug) failure: test compiler-rt debug (failure)
> ...
> [191/238] Linking CXX static library lib/libclangFrontend.a
> [192/238] Linking CXX static library lib/libclangIndex.a
> [193/238] Linking CXX static library lib/libclangRewriteFrontend.a
> [194/238] Linking CXX static library lib/libclangARCMigrate.a
> [195/238] Linking CXX static library lib/libclangCrossTU.a
> [196/238] Linking CXX static library lib/libclangExtractAPI.a
> [197/238] Linking CXX static library lib/libclangStaticAnalyzerCore.a
> [198/238] Linking CXX static library lib/libclangStaticAnalyzerCheckers.a
> [199/238] Linking CXX static library lib/libclangStaticAnalyzerFrontend.a
> [200/238] Building CXX object lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o
> FAILED: lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o 
> CCACHE_CPP2=yes CCACHE_HASHDIR=yes /usr/bin/ccache /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm_build0/bin/clang++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -MF lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o.d -o lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -c /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp
> In file included from /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:20:
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizer.h:36:28: error: private field 'Options' is not used [-Werror,-Wunused-private-field]
>    36 |   RealtimeSanitizerOptions Options{};
>       |                            ^
> 1 error generated.
> ninja: build stopped: subcommand failed.
> + build_failure
> + echo
> + echo 'How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild'
> 
> How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild
> + echo
> + sleep 5
> + echo @@@STEP_FAILURE@@@
> + [[ '' == \1 ]]
> + [[ ! -v BUILDBOT_BUILDERNAME ]]
> + build tsan_debug '-DCOMPILER_RT_DEBUG=ON -DCOMPILER_RT_TSAN_DEBUG_OUTPUT=ON -DLLVM_INCLUDE_TESTS=OFF -DCOMPILER_RT_BUILD_LIBFUZZER=OFF'
> + [[  -DLLVM_APPEND_VC_REV=OFF -GNinja -DCMAKE_BUILD_TYPE=Release -DLLVM_CCACHE_BUILD=ON -DLLVM_USE_LINKER=lld -DLLVM_TARGETS_TO_BUILD=PowerPC -DLLVM_LIT_ARGS=-vj256 -DLLVM_ENABLE_PROJECTS=clang;lld -DLLVM_ENABLE_RUNTIMES=libcxx;libcxxabi;compiler-rt;libunwind -DLLVM_BUILD_LLVM_DYLIB=ON -DLLVM_ENABLE_ASSERTIONS=ON  -DCMAKE_C_COMPILER=/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm_build0/bin/clang -DCMAKE_CXX_COMPILER=/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm_build0/bin/clang++ -DLLVM_ENABLE_WERROR=ON =~ LLVM_CCACHE_BUILD=ON ]]
> + BUILD_DIR=build_default
> + rm -rf build_default
> + echo '@@@BUILD_STEP build compiler-rt tsan_debug@@@'
> + [[ ! -f build_default/delete_next_time ]]
> + mkdir -p build_default
> Step 10 (build compiler-rt tsan_debug) failure: build compiler-rt tsan_debug (failure)
> ...
> [3624/3988] Linking CXX executable bin/llvm-bcanalyzer
> [3625/3988] Linking CXX executable bin/llvm-dis
> [3626/3988] Building CXX object tools/clang/tools/clang-check/CMakeFiles/clang-check.dir/ClangCheck.cpp.o
> [3627/3988] Building CXX object tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndex.cpp.o
> [3628/3988] Building CXX object tools/clang/tools/clang-repl/CMakeFiles/clang-repl.dir/ClangRepl.cpp.o
> [3629/3988] Building CXX object tools/clang/tools/clang-scan-deps/CMakeFiles/clang-scan-deps.dir/ClangScanDeps.cpp.o
> [3630/3988] Linking CXX executable bin/llvm-diff
> [3631/3988] Building CXX object tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXExtractAPI.cpp.o
> [3632/3988] Building CXX object tools/clang/tools/clang-installapi/CMakeFiles/clang-installapi.dir/ClangInstallAPI.cpp.o
> [3633/3988] Building CXX object lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o
> FAILED: lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o 
> CCACHE_CPP2=yes CCACHE_HASHDIR=yes /usr/bin/ccache /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm_build0/bin/clang++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -MF lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o.d -o lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -c /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp
> In file included from /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:20:
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizer.h:36:28: error: private field 'Options' is not used [-Werror,-Wunused-private-field]
>    36 |   RealtimeSanitizerOptions Options{};
>       |                            ^
> 1 error generated.
> [3634/3988] Building CXX object tools/clang/lib/Driver/CMakeFiles/obj.clangDriver.dir/ToolChains/HIPUtility.cpp.o
> [3635/3988] Building CXX object lib/MC/MCParser/CMakeFiles/LLVMMCParser.dir/AsmParser.cpp.o
> ninja: build stopped: subcommand failed.
> + touch build_default/delete_next_time
> + build_failure
> + echo
> + echo 'How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild'
> + echo
> + sleep 5
> 
> How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild
> + echo @@@STEP_FAILURE@@@
> + [[ '' == \1 ]]
> + [[ ! -v BUILDBOT_BUILDERNAME ]]
> + build_and_test default ''
> + build default ''
> + [[  -DLLVM_APPEND_VC_REV=OFF -GNinja -DCMAKE_BUILD_TYPE=Release -DLLVM_CCACHE_BUILD=ON -DLLVM_USE_LINKER=lld -DLLVM_TARGETS_TO_BUILD=PowerPC -DLLVM_LIT_ARGS=-vj256 -DLLVM_ENABLE_PROJECTS=clang;lld -DLLVM_ENABLE_RUNTIMES=libcxx;libcxxabi;compiler-rt;libunwind -DLLVM_BUILD_LLVM_DYLIB=ON -DLLVM_ENABLE_ASSERTIONS=ON  -DCMAKE_C_COMPILER=/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm_build0/bin/clang -DCMAKE_CXX_COMPILER=/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm_build0/bin/clang++ -DLLVM_ENABLE_WERROR=ON =~ LLVM_CCACHE_BUILD=ON ]]
> + BUILD_DIR=build_default
> + rm -rf build_default
> + echo '@@@BUILD_STEP build compiler-rt default@@@'
> + [[ ! -f build_default/delete_next_time ]]
> + mkdir -p build_default
> Step 11 (build compiler-rt default) failure: build compiler-rt default (failure)
> ...
> [3643/4007] Linking CXX executable bin/llvm-diff
> [3644/4007] Building CXX object tools/clang/tools/driver/CMakeFiles/clang.dir/driver.cpp.o
> [3645/4007] Building CXX object tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndex.cpp.o
> [3646/4007] Building CXX object tools/clang/tools/c-index-test/CMakeFiles/c-index-test.dir/core_main.cpp.o
> [3647/4007] Building CXX object tools/clang/tools/clang-import-test/CMakeFiles/clang-import-test.dir/clang-import-test.cpp.o
> [3648/4007] Building CXX object tools/clang/tools/clang-repl/CMakeFiles/clang-repl.dir/ClangRepl.cpp.o
> [3649/4007] Building CXX object tools/clang/tools/clang-scan-deps/CMakeFiles/clang-scan-deps.dir/ClangScanDeps.cpp.o
> [3650/4007] Building CXX object tools/clang/tools/clang-installapi/CMakeFiles/clang-installapi.dir/ClangInstallAPI.cpp.o
> [3651/4007] Building CXX object tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXExtractAPI.cpp.o
> [3652/4007] Building CXX object lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o
> FAILED: lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o 
> CCACHE_CPP2=yes CCACHE_HASHDIR=yes /usr/bin/ccache /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm_build0/bin/clang++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -MF lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o.d -o lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -c /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp
> In file included from /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:20:
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizer.h:36:28: error: private field 'Options' is not used [-Werror,-Wunused-private-field]
>    36 |   RealtimeSanitizerOptions Options{};
>       |                            ^
> 1 error generated.
> [3653/4007] Building CXX object tools/clang/lib/Driver/CMakeFiles/obj.clangDriver.dir/ToolChains/HIPUtility.cpp.o
> [3654/4007] Building CXX object lib/MC/MCParser/CMakeFiles/LLVMMCParser.dir/AsmParser.cpp.o
> ninja: build stopped: subcommand failed.
> + touch build_default/delete_next_time
> + build_failure
> + echo
> + echo 'How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild'
> + echo
> + sleep 5
> 
> How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild
> + echo @@@STEP_FAILURE@@@
> + [[ '' == \1 ]]
> + [[ ! -v BUILDBOT_BUILDERNAME ]]
> + echo '@@@BUILD_STEP test compiler-rt default@@@'
> Step 12 (test compiler-rt default) failure: test compiler-rt default (failure)
> ...
> [191/238] Linking CXX static library lib/libclangFrontend.a
> [192/238] Linking CXX static library lib/libclangIndex.a
> [193/238] Linking CXX static library lib/libclangRewriteFrontend.a
> [194/238] Linking CXX static library lib/libclangARCMigrate.a
> [195/238] Linking CXX static library lib/libclangCrossTU.a
> [196/238] Linking CXX static library lib/libclangExtractAPI.a
> [197/238] Linking CXX static library lib/libclangStaticAnalyzerCore.a
> [198/238] Linking CXX static library lib/libclangStaticAnalyzerCheckers.a
> [199/238] Linking CXX static library lib/libclangStaticAnalyzerFrontend.a
> [200/238] Building CXX object lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o
> FAILED: lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o 
> CCACHE_CPP2=yes CCACHE_HASHDIR=yes /usr/bin/ccache /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm_build0/bin/clang++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -MF lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o.d -o lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/RealtimeSanitizer.cpp.o -c /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp
> In file included from /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:20:
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizer.h:36:28: error: private field 'Options' is not used [-Werror,-Wunused-private-field]
>    36 |   RealtimeSanitizerOptions Options{};
>       |                            ^
> 1 error generated.
> ninja: build stopped: subcommand failed.
> + build_failure
> + echo
> 
> How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild
> + echo 'How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild'
> + echo
> + sleep 5
> + echo @@@STEP_FAILURE@@@
> + [[ '' == \1 ]]
> + [[ ! -v BUILDBOT_BUILDERNAME ]]
> + FRESH_CLANG_PATH=/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/bin
> + echo @@@BUILD_STEP build standalone compiler-rt@@@
> + mkdir -p compiler_rt_build
> Step 13 (build standalone compiler-rt) failure: build standalone compiler-rt (failure)
> ...
>   Tell CMake where to find the compiler by setting either the environment
>   variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
>   the compiler, or to the compiler name if it is in the PATH.
> 
> 
> CMake Error at CMakeLists.txt:17 (project):
>   The CMAKE_CXX_COMPILER:
> 
>     /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/build_default/bin/clang++
> 
>   is not a full path to an existing compiler tool.
> 
>   Tell CMake where to find the compiler by setting either the environment
>   variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
>   to the compiler, or to the compiler name if it is in the PATH.
> 
> 
> CMake Error at CMakeLists.txt:17 (project):
>   No CMAKE_ASM_COMPILER could be found.
> 
>   Tell CMake where to find the compiler by setting either the environment
>   variable "ASM" or the CMake cache entry CMAKE_ASM_COMPILER to the full path
>   to the compiler, or to the compiler name if it is in the PATH.
> -- Warning: Did not find file Compiler/-ASM
> -- Configuring incomplete, errors occurred!
> + build_failure
> + echo
> + echo 'How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild'
> 
> How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild
> + echo
> + sleep 5
> + echo @@@STEP_FAILURE@@@
> + [[ '' == \1 ]]
> + [[ ! -v BUILDBOT_BUILDERNAME ]]
> + ninja -C compiler_rt_build
> ninja: Entering directory `compiler_rt_build'
> ninja: error: loading 'build.ninja': No such file or directory
> + build_failure
> + echo
> 
> + echo 'How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild'
> + echo
> + sleep 5
> How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild
> + echo @@@STEP_FAILURE@@@
> + [[ '' == \1 ]]
> + [[ ! -v BUILDBOT_BUILDERNAME ]]
> + echo @@@BUILD_STEP test standalone compiler-rt@@@
> + ninja -C compiler_rt_build check-all
> Step 14 (test standalone compiler-rt) failure: test standalone compiler-rt (failure)
> @@@BUILD_STEP test standalone compiler-rt@@@
> ninja: Entering directory `compiler_rt_build'
> ninja: error: loading 'build.ninja': No such file or directory
> + build_failure
> + echo
> 
> + echo 'How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild'
> + echo
> + sleep 5
> How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild
> + echo @@@STEP_FAILURE@@@
> + [[ '' == \1 ]]
> + [[ ! -v BUILDBOT_BUILDERNAME ]]
> + '[' 0 == 1 ']'
> + cleanup
> + [[ -v BUILDBOT_BUILDERNAME ]]
> + echo @@@BUILD_STEP cleanup@@@
> ```

Fixed in f86594788ce93b696675c94f54016d27a6c21d18 

Therefore, we recommend that unused variables are not added in a patch, even if it will soon be used by the next patch.

cjappl commented:
> Therefore, we recommend that unused variables are not added in a patch, even if it will soon be used by the next patch.

That makes sense!! Thanks for the speedy patch

cjappl commented:
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/145/builds/1104

Looking at the build log of this last one, it appears to be the same problem with the private member. This is fixed in main

nikic commented on llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:
```diff
@@ -0,0 +1,61 @@
+  if (F.hasFnAttribute(Attribute::SanitizeRealtime)) {
+    insertCallAtFunctionEntryPoint(F, "__rtsan_realtime_enter");
+    insertCallAtAllFunctionExitPoints(F, "__rtsan_realtime_exit");
+    return PreservedAnalyses::none();
```
Not super important, but this could `preserveSet<CFGAnalyses>`.

cjappl replied:
Thanks! I can fix this in a follow up.

Do you know where I can find more info on what this means? I was pattern matching other transforms and don't have a deep knowledge of what these "preserved analyses" actually do for the pass.

Any additional info greatly appreciated!

fmayer replied:
It means whether the analysis has to be re-evalulated after the pass (if needed by a pass afterwards), or whether the pass has not changed the result (or, for some analyses, it is possible to incrementally change them, in which case it could also have done that).

cjappl replied:
Ok, if I can rephrase what you're saying for my own education:

"If a pass has changed the code in some way, we need to alert the analysis that it has been changed so it can be re-evaluated.

In my original code, I basically said the maximal 'this has changed EVERYTHING, please re-run EVERYTHING (preserve nothing)'. Where if I was being more judicious, I would have only said 'I changed the code, but the Control Flow Analysis is still valid, you do not have to re-run that portion, it is preserved'
"

Is my understanding correct? 

Sincerely appreciate the information here, this area of the project is brand new to me and I will take any knowledge I can get.

fmayer replied:
Yes, that sounds correct.
